### PR TITLE
nm.bridge: get ports from sysfs

### PR DESCRIPTION
**Pull request by ffmancera**
Unmanaged ports were not being shown when showing or editting the
bridge. In order to fix that, nmstate is retrieving the bridge ports
from sysfs instead of NetworkManager on-disk configuration.

Ref: https://bugzilla.redhat.com/show_bug.cgi?id=1806452

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

### CODE REVIEW

**Comment by cathay4t** [comment]
@ffmancera Can I put this PR on hold tile PR #576 related stuff merged?
I don't want to mess with outside contributors' work.

**Comment by ffmancera** [comment]
> @ffmancera Can I put this PR on hold till PR #576 related stuff merged?
> I don't want to mess with outside contributors' work.

Sure, no worries.

**Comment by ffmancera** [comment]
Waiting for #576.

**Comment by tyll** [comment]
There is already a conflict that could be resolved

File: tests/integration/linux_bridge_test.py [hunk 1]
```diff
@@ -410,6 +412,32 @@ def test_port_vlan_not_implemented(port0_up):
...
+            port_subtree = bridge_state[LinuxBridge.CONFIG_SUBTREE].get(
+                LinuxBridge.PORT_SUBTREE, []
+            )
+
+            assert len(port_subtree) == 1
+
+
+@contextmanager
+def dummy0_as_slave(master):
+    exec_cmd(("ip", "link", "add", "dummy0", "type", "dummy"))
```
Comment (cathay4t) [inline tests/integration/linux_bridge_test.py]: `check=True` is required.

File: tests/integration/nm/linux_bridge_test.py [hunk 2]
```diff
@@ -64,16 +63,6 @@ def test_create_and_remove_bridge(port0_up):
     assert not _get_bridge_current_state()
 
 
-@iproutelib.ip_monitor_assert_stable_link_up(BRIDGE0)
-def test_add_port_to_existing_bridge(bridge0_with_port0, port1_up):
```
Comment (cathay4t) [inline tests/integration/nm/linux_bridge_test.py]: Why this test case been removed.

Comment (ffmancera) [reply tests/integration/nm/linux_bridge_test.py]: That was wrong. Fixed.

File: tests/integration/linux_bridge_test.py (hunk 1 above)
Comment (ffmancera) [reply tests/integration/linux_bridge_test.py]: Done, thanks!

File: libnmstate/nm/bridge.py [hunk 3]
```diff
@@ -165,36 +176,56 @@ def _get_bridge_setting(nmdev):
...
+def _get_slaves_runtime(master):
+    """
+    We need to use glob in order to get the slaves name due to bug in
+    NetworkManager.
+    Ref: https://bugzilla.redhat.com/show_bug.cgi?id=1809547
+    """
+    slaves = []
+    for sysfs_slave in glob.iglob(f"/sys/class/net/{master}/lower_*"):
+        # The format is lower_<iface>, we need to remove the "lower_" prefix
+        slaves.append(os.path.basename(sysfs_slave)[6:])
```
Comment (cathay4t) [inline libnmstate/nm/bridge.py]: The `6` should be `len("lower_")`.

File: tests/integration/linux_bridge_test.py [hunk 4]
```diff
@@ -537,6 +539,37 @@ def test_disable_vlan_filtering_on_bridge(
...
+    bridge_config = _create_bridge_subtree_config([])
+    with linux_bridge(TEST_BRIDGE0, bridge_config):
+        with dummy0_as_slave(TEST_BRIDGE0):
+            current_state = show_only((TEST_BRIDGE0,))
+            bridge_state = current_state[Interface.KEY][0]
+            port_subtree = bridge_state[LinuxBridge.CONFIG_SUBTREE].get(
+                LinuxBridge.PORT_SUBTREE, []
+            )
+
+            assert len(port_subtree) == 1
```
Comment (cathay4t) [inline tests/integration/linux_bridge_test.py]: Please assert the content of `port_subtree` instead of length.

File: libnmstate/nm/bridge.py (hunk 3 above)
Comment (ffmancera) [reply libnmstate/nm/bridge.py]: Done, thanks!

File: tests/integration/linux_bridge_test.py (hunk 4 above)
Comment (ffmancera) [reply tests/integration/linux_bridge_test.py]: Done, thanks!

File: libnmstate/nm/bridge.py [hunk 5]
```diff
@@ -165,36 +176,57 @@ def _get_bridge_setting(nmdev):
...
-    ports_info = []
-    for p in port_profiles:
-        port_info = _get_bridge_port_info(p)
+def _get_bridge_ports_info(
+    port_names_profiles, port_runtime_names, vlan_filtering_enabled=False
+):
+    ports_name_info = {}
+    for name in port_runtime_names:
+        port_info = _get_bridge_port_runtime_config(name)
+        ports_name_info[name] = port_info
```
Comment (EdDev) [inline libnmstate/nm/bridge.py]: Seems to fit a dict comprehension.

File: libnmstate/nm/bridge.py [hunk 6]
```diff
@@ -165,36 +176,57 @@ def _get_bridge_setting(nmdev):
...
 
 
-def _get_bridge_ports_info(port_profiles, vlan_filtering_enabled=False):
-    ports_info = []
-    for p in port_profiles:
-        port_info = _get_bridge_port_info(p)
+def _get_bridge_ports_info(
+    port_names_profiles, port_runtime_names, vlan_filtering_enabled=False
+):
+    ports_name_info = {}
```
Comment (EdDev) [inline libnmstate/nm/bridge.py]: I do not understand the name.. it seems to be `ports_info_by_name`.

File: libnmstate/nm/bridge.py [hunk 7]
```diff
@@ -165,36 +176,57 @@ def _get_bridge_setting(nmdev):
...
-    """Report port information."""
-
-    port_setting = port_profile.get_setting_bridge_port()
-    return {
-        LB.Port.NAME: port_profile.get_interface_name(),
-        LB.Port.STP_PRIORITY: port_setting.props.priority,
-        LB.Port.STP_HAIRPIN_MODE: port_setting.props.hairpin_mode,
-        LB.Port.STP_PATH_COST: port_setting.props.path_cost,
-    }
+    return [port_info for port_info in ports_name_info.values()]
```
Comment (EdDev) [inline libnmstate/nm/bridge.py]: Isn't this `list(ports_name_info.values())` ?

File: libnmstate/nm/bridge.py [hunk 8]
```diff
@@ -165,36 +176,57 @@ def _get_bridge_setting(nmdev):
...
-    slave_profiles = []
+def _get_slave_names_profiles(master_device):
+    slaves_names_profiles = []
     for dev in master_device.get_slaves():
         active_con = connection.get_device_active_connection(dev)
         if active_con:
-            slave_profiles.append(active_con.props.connection)
-    return slave_profiles
+            slave_name_profile = (dev.get_iface(), active_con.props.connection)
+            slaves_names_profiles.append(slave_name_profile)
```
Comment (EdDev) [inline libnmstate/nm/bridge.py]: This is ugly IMO. Can you make this a dict (i.e. `slaves_profiles_by_name`) ?

File: libnmstate/nm/bridge.py [hunk 9]
```diff
@@ -165,36 +176,57 @@ def _get_bridge_setting(nmdev):
...
+    return slaves_names_profiles
+
+
+def _get_bridge_port_runtime_config(port_runtime):
```
Comment (EdDev) [inline libnmstate/nm/bridge.py]: `config` -> `info`
`port_runtime` -> `port_name`

File: libnmstate/nm/bridge.py [hunk 10]
```diff
@@ -165,36 +176,57 @@ def _get_bridge_setting(nmdev):
...
+    """Report port runtime information from sysfs."""
+    port = {LB.Port.NAME: port_runtime}
+    for option, option_sysfs in BRIDGE_PORT_NMSTATE_TO_SYSFS.items():
+        sysfs_path = f"/sys/class/net/{port_runtime}/brport/{option_sysfs}"
+        with open(sysfs_path) as f:
+            option_value = int(f.read())
+            if option == LB.Port.STP_HAIRPIN_MODE:
+                option_value = bool(option_value)
+            port[option] = option_value
```
Comment (EdDev) [inline libnmstate/nm/bridge.py]: This does not need to be inside the `open` block.

File: tests/integration/linux_bridge_test.py [hunk 11]
```diff
@@ -537,6 +539,37 @@ def test_disable_vlan_filtering_on_bridge(
...
+    bridge_config = _create_bridge_subtree_config([])
+    with linux_bridge(TEST_BRIDGE0, bridge_config):
+        with dummy0_as_slave(TEST_BRIDGE0):
+            current_state = show_only((TEST_BRIDGE0,))
+            bridge_state = current_state[Interface.KEY][0]
+            port_subtree = bridge_state[LinuxBridge.CONFIG_SUBTREE].get(
+                LinuxBridge.PORT_SUBTREE, []
+            )
+
+            assert port_subtree[0].get(LinuxBridge.Port.NAME) == "dummy0"
```
Comment (EdDev) [inline tests/integration/linux_bridge_test.py]: There are some entries you assume that exists and some that do not.
Lets assume all exist, otherwise it should fail anyway (an alternative is to assert all the time).

What I mean is not to use `get` and its defaults.

File: tests/integration/linux_bridge_test.py [hunk 12]
```diff
@@ -537,6 +539,37 @@ def test_disable_vlan_filtering_on_bridge(
...
+
+
+@contextmanager
+def dummy0_as_slave(master):
+    try:
+        exec_cmd(("ip", "link", "add", "dummy0", "type", "dummy"), check=True)
```
Comment (EdDev) [inline tests/integration/linux_bridge_test.py]: The `try` block should start *after* this line.

File: tests/integration/linux_bridge_test.py [hunk 13]
```diff
@@ -537,6 +539,37 @@ def test_disable_vlan_filtering_on_bridge(
...
+        exec_cmd(("ip", "link", "set", "dummy0", "up"), check=True)
+        exec_cmd(
+            ("nmcli", "dev", "set", "dummy0", "managed", "no"), check=True
+        )
+        exec_cmd(("ip", "link", "set", "dummy0", "master", master), check=True)
+    except Exception:
+        exec_cmd(("ip", "link", "delete", "dummy0"))
+
```
Comment (EdDev) [inline tests/integration/linux_bridge_test.py]: You need to raise here otherwise it will continue.
But I do not think you need this `Exception` block at all. If it fails, just get to the `finally` and that is it. One by try/finally block fits here, not two.

File: tests/integration/nm/linux_bridge_test.py [hunk 14]
```diff
@@ -154,7 +154,12 @@ def _modify_ports(ports_state):
 @nmclient_context
 def _get_bridge_current_state():
     nmdev = nm.device.get_device_by_name(BRIDGE0)
-    return nm.bridge.get_info(nmdev) if nmdev else {}
+    state = nm.bridge.get_info(nmdev) if nmdev else {}
+    if state:
+        slaves = state.get(LB.CONFIG_SUBTREE, {}).get(LB.PORT_SUBTREE, [])
+        slaves.sort(key=lambda d: d[LB.Port.NAME])
```
Comment (EdDev) [inline tests/integration/nm/linux_bridge_test.py]: This seems to be an unrelated fix. Can you extract it to its own commit? (with an explanation in the commit message)

File: tests/integration/linux_bridge_test.py [hunk 15]
```diff
@@ -537,6 +539,37 @@ def test_disable_vlan_filtering_on_bridge(
             assertlib.assert_state_match(desired_state)
 
 
+def test_bridge_with_unmanaged_ports():
+    bridge_config = _create_bridge_subtree_config([])
+    with linux_bridge(TEST_BRIDGE0, bridge_config):
+        with dummy0_as_slave(TEST_BRIDGE0):
```
Comment (EdDev) [inline tests/integration/linux_bridge_test.py]: The test itself is about detecting unmanaged slaves when reporting the state.
Therefore, the test setup belongs to a fixture IMO.

File: libnmstate/nm/bridge.py (hunk 5 above)
Comment (ffmancera) [reply libnmstate/nm/bridge.py]: Done, thanks!

File: libnmstate/nm/bridge.py (hunk 6 above)
Comment (ffmancera) [reply libnmstate/nm/bridge.py]: Done, thanks!

File: libnmstate/nm/bridge.py (hunk 7 above)
Comment (ffmancera) [reply libnmstate/nm/bridge.py]: Omg.. yes. Thank you!

File: libnmstate/nm/bridge.py (hunk 8 above)
Comment (ffmancera) [reply libnmstate/nm/bridge.py]: Yes, done thanks!

File: libnmstate/nm/bridge.py (hunk 9 above)
Comment (ffmancera) [reply libnmstate/nm/bridge.py]: Done, thanks!

File: libnmstate/nm/bridge.py (hunk 10 above)
Comment (ffmancera) [reply libnmstate/nm/bridge.py]: Oops. Done!

File: tests/integration/linux_bridge_test.py (hunk 11 above)
Comment (ffmancera) [reply tests/integration/linux_bridge_test.py]: Done, thanks!

File: tests/integration/linux_bridge_test.py (hunk 13 above)
Comment (ffmancera) [reply tests/integration/linux_bridge_test.py]: If some of the commands fail then there is going to be a leftover. I am raising after removing the interface.

File: tests/integration/linux_bridge_test.py (hunk 12 above)
Comment (ffmancera) [reply tests/integration/linux_bridge_test.py]: Yes, you are right. Done.

File: tests/integration/linux_bridge_test.py (hunk 15 above)
Comment (ffmancera) [reply tests/integration/linux_bridge_test.py]: Done, thanks!

File: tests/integration/nm/linux_bridge_test.py (hunk 14 above)
Comment (ffmancera) [reply tests/integration/nm/linux_bridge_test.py]: Done, thanks!

File: libnmstate/nm/bridge.py [hunk 16]
```diff
@@ -165,36 +176,57 @@ def _get_bridge_setting(nmdev):
...
-        LB.Port.NAME: port_profile.get_interface_name(),
-        LB.Port.STP_PRIORITY: port_setting.props.priority,
-        LB.Port.STP_HAIRPIN_MODE: port_setting.props.hairpin_mode,
-        LB.Port.STP_PATH_COST: port_setting.props.path_cost,
-    }
-
-
-def _get_slave_profiles(master_device):
-    slave_profiles = []
+def _get_slave_names_profiles(master_device):
```
Comment (EdDev) [inline libnmstate/nm/bridge.py]: Can you sync with the variable name?

File: libnmstate/nm/bridge.py [hunk 17]
```diff
@@ -165,36 +176,57 @@ def _get_bridge_setting(nmdev):
...
         if active_con:
-            slave_profiles.append(active_con.props.connection)
-    return slave_profiles
+            slaves_profiles_by_name[
+                dev.get_iface()
+            ] = active_con.props.connection
+    return slaves_profiles_by_name
+
+
+def _get_bridge_port_name_info(port_runtime):
```
Comment (EdDev) [inline libnmstate/nm/bridge.py]: How about `_get_bridge_port_info(port_name):` ?

File: libnmstate/nm/bridge.py [hunk 18]
```diff
@@ -165,36 +176,57 @@ def _get_bridge_setting(nmdev):
...
+        sysfs_path = f"/sys/class/net/{port_runtime}/brport/{option_sysfs}"
+        with open(sysfs_path) as f:
+            option_value = int(f.read())
+            if option == LB.Port.STP_HAIRPIN_MODE:
+                option_value = bool(option_value)
+        port[option] = option_value
+    return port
+
+
+def _get_slaves_runtime(master):
```
Comment (EdDev) [inline libnmstate/nm/bridge.py]: How about `_get_slaves_names_from_runtime`?
`runtime` can be changed to `sysfs`.

File: tests/integration/linux_bridge_test.py (hunk 13 above)
Comment (EdDev) [reply tests/integration/linux_bridge_test.py]: You do this in the finally, that block is called for both regular exit and exceptions alike.

Comment (ffmancera) [reply tests/integration/linux_bridge_test.py]: Done, thanks!

File: libnmstate/nm/bridge.py (hunk 16 above)
Comment (ffmancera) [reply libnmstate/nm/bridge.py]: Done, thank you!

File: libnmstate/nm/bridge.py (hunk 17 above)
Comment (ffmancera) [reply libnmstate/nm/bridge.py]: Done!

File: libnmstate/nm/bridge.py (hunk 18 above)
Comment (ffmancera) [reply libnmstate/nm/bridge.py]: Done! Thanks!

**Comment by ffmancera** [comment]
Rebased! Thanks!
